controller/appcontroller.go

```diff
@@ -2110,18 +2116,45 @@ func alreadyAttemptedSync(app *appv1.Application, commitSHA string, commitSHAsMS
 	}
 }
 
-func (ctrl *ApplicationController) shouldSelfHeal(app *appv1.Application) (bool, time.Duration) {
+func (ctrl *ApplicationController) shouldSelfHeal(app *appv1.Application) (bool, time.Duration, time.Duration) {
 	if app.Status.OperationState == nil {
-		return true, time.Duration(0)
+		return true, time.Duration(0), time.Duration(0)
 	}
-
 	var retryAfter time.Duration
+	timeSinceLastOperationEnd := time.Since(app.Status.OperationState.FinishedAt.Time)
+
+	nextSelfHealBackoffTimeout := ctrl.nextSelfHealBackoffTimeout(app)
 	if app.Status.OperationState.FinishedAt == nil {
-		retryAfter = ctrl.selfHealTimeout
+		retryAfter = nextSelfHealBackoffTimeout
 	} else {
-		retryAfter = ctrl.selfHealTimeout - time.Since(app.Status.OperationState.FinishedAt.Time)
+		retryAfter = nextSelfHealBackoffTimeout - timeSinceLastOperationEnd
+	}
+	return retryAfter <= 0, retryAfter, nextSelfHealBackoffTimeout
+}
+
+// isSelfHealBackoffEnabled returns whether self heal exponential backoff
+// is enabled.
+func (ctrl *ApplicationController) isSelfHealBackoffEnabled(app *appv1.Application) bool {
+	return len(app.Status.History) > 0 && ctrl.selfHealTimeoutBackoffFactor > 0 && ctrl.selfHealTimeoutBackoffMax > 0
+}
+
+// nextSelfHealBackoffTimeout returns the next self heal timeout
+func (ctrl *ApplicationController) nextSelfHealBackoffTimeout(app *appv1.Application) time.Duration {
+
+	if ctrl.isSelfHealBackoffEnabled(app) {
+		latestDeployTime := app.Status.History.LastRevisionHistory().DeployedAt.Time
```

Comment:
This is the closest time I could identify to represent T0 in the backoff timeline. Open to any suggestions for improvement.